AX: Implement TextUnit::LeftWord/RightWord off the main thread

hoffmanjoshua opened the pull request:
#### bcf367428ee3db2103c809b74980932a50f902f2
<pre>
AX: Implement TextUnit::LeftWord/RightWord off the main thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=280825">https://bugs.webkit.org/show_bug.cgi?id=280825</a>
<a href="https://rdar.apple.com/134579687">rdar://134579687</a>

Reviewed by Tyler Wilcock.

This patch implements the `AXLeftWordTextMarkerRangeForTextMarker` and `AXRightWordTextMarkerRangeForTextMarker` attributes using the secondary thread text markers. To mimic the live tree’s current behavior, several edges cases had to be considered:
* If at the end of a paragraph (containing block), return without a word.
* If we are finding the left word, and we are at a start word boundary, that should return the whitespace prior to the word (if any).
* If we are finding the right word, and are at the end boundary of a word, return the whitespace to the right of the word.

To verify this new behavior, the test `text-marker-word-nav` was ported over from the live tree, with some modifications. APIs that still used the live tree were removed, due to some issues with interop between live and isolated tree text markers. Three test cases were are currently failing with FIXMEs: two due to the word break iterator with Chinese characters, and one due to line break behavior.

In order to accurately accommodate handling word navigation, I had to make a change to the `findMarker` method to allow for coalescing object boundaries. This is the cases where the end position of one object is equivalent visually/in position to the start position of the next object. Coalescing mimics behavior from the live tree.

* LayoutTests/accessibility/ax-thread-text-apis/text-marker-word-nav-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/text-marker-word-nav.html: Added.
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::lineNumberForIndex const):
(WebCore::AXTextMarker::atLineBoundaryForDirection const):
(WebCore::AXTextMarker::offsetFromRoot const):
(WebCore::AXTextMarker::findLastBefore const):
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::wordRange const):
(WebCore::AXTextMarker::equivalentTextPosition const):
(WebCore::AXTextMarker::partialOrderByTraversal const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::nextWordStart const):
(WebCore::AXTextMarker::nextWordEnd const):
(WebCore::AXTextMarker::previousWordStart const):
(WebCore::AXTextMarker::previousWordEnd const):
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRuns::toString const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
(WebCore::AccessibilityRenderObject::textRuns):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper textMarkerRangeAtTextMarker:forUnit:]):

Canonical link: <a href="https://commits.webkit.org/284938@main">https://commits.webkit.org/284938@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed27371c033acf7c03262d284804184f7922bcdb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70940 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50352 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23713 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75044 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22145 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73056 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58151 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21963 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14595 "1 failures") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74006 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45761 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61159 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36563 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/42416 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/18593 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20486 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64356 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18954 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76759 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15172 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/18154 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63856 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15216 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61218 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63812 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15709 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11905 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/5552 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46153 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/928 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47225 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48508 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46967 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/ax-thread-text-apis/text-marker-word-nav.html:
```diff
@@ -0,0 +1,191 @@
+    // FIXME: Previous expectation had left and right word both be empty. This new expectation expects that the right and left words are the words on either side of the line break.
+    currentMarker = advanceAndVerify(currentMarker, 1, text6);
+
+    // FIXME: Skipping text 8 from the original text (which tests <audio> on the live tree), due to invalidation issues with <audio controls> and the AXObjectCache.
```
I'm curious (and would be nice to include in the comment if you remember) — what are the details of this invalidation issue? Is it an ITM cache invalidation issue?

hoffmanjoshua replied:
This was due to `gAccessibilityThreadTextApis` being set to the wrong value by another document. This was fixed with https://github.com/WebKit/WebKit/commit/18aeb0e661445201b72923ce20216bc52033b2a3

twilco commented on LayoutTests/accessibility/ax-thread-text-apis/text-marker-word-nav.html:
```diff
@@ -0,0 +1,191 @@
+    currentMarker = advanceAndVerify(currentMarker, 1, text4);
+    currentMarker = advanceAndVerify(currentMarker, 1, text4);
+
+    // // And Arabic
```
Double // here

twilco commented on LayoutTests/accessibility/ax-thread-text-apis/text-marker-word-nav.html:
```diff
@@ -0,0 +1,191 @@
+    textMarkerRange = text4.textMarkerRangeForElement(text4);
+    currentMarker = text4.startTextMarkerForTextMarkerRange(textMarkerRange);
+    // Make sure when we are at the beginning of line, it won't go to previous node.
+    // FIXME: Left word in the live tree returns a word, which is inconsistent with other left words when at the beginning of a word.
```
Do we know why the main-thread text marker implementation behaves differently / inconsistently in this case?

hoffmanjoshua replied:
I haven't been able to nail this down precisely. The main thread/visible position has a lot of code to handle stopping at word boundaries. `<br>` characters are handled very differently in several scenarios.

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -732,20 +740,40 @@ const AXTextRuns* AXTextMarker::runs() const
 }
 
-AXTextMarker AXTextMarker::findMarker(AXDirection direction, std::optional<AXID> stopAtID) const
+AXTextMarker AXTextMarker::findMarker(AXDirection direction, bool coalesceObjectBreaks, bool ignoreBRs, std::optional<AXID> stopAtID) const
```
Rather than `bool`s, could we make these enum classes? e.g. `enum class CoalesceObjectBreaks : bool { No, Yes }`

This would help make the callsites a bit more descriptive vs. `false` / `true`.

hoffmanjoshua replied:
Yes, great idea!

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -732,20 +740,40 @@ const AXTextRuns* AXTextMarker::runs() const
+AXTextMarker AXTextMarker::findMarker(AXDirection direction, bool coalesceObjectBreaks, bool ignoreBRs, std::optional<AXID> stopAtID) const
 {
+    // This method has two boolean options:
+    // - coalesceObjectBreaks: Mimics behavior from textMarkerDataForNextCharacterOffset, where we skip nodes
+    //   that have the same visual position (i.e., there is 0 length between them). When false, we traverse all
+    //   possible text markers (which is important for searching)
+    // - ignoreBRs: In most cases, we want to skip <br> tags when not in an editable context. This is not true,
+    //   for example, when computing text marker indexes.
```
Nice documentation on these!

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -732,20 +740,40 @@ const AXTextRuns* AXTextMarker::runs() const
+    RefPtr object = isolatedObject();
+
+    // If the BR isn't in an editable ancestor, we shouldn't be including it (in most cases of findMarker).
+    bool shouldSkipBR = (ignoreBRs && !object->editableAncestor()) && object && object->roleValue() == AccessibilityRole::LineBreak;
```
`object && object->roleValue() == AccessibilityRole::LineBreak` is significantly cheaper of a check vs. `editableAncestor()` which does an ancestry traversal, so let's put that part first. This ensures we only do the expensive work when we really have to (for line breaks).

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -732,20 +740,40 @@ const AXTextRuns* AXTextMarker::runs() const
+    bool shouldSkipBR = (ignoreBRs && !object->editableAncestor()) && object && object->roleValue() == AccessibilityRole::LineBreak;
+    if (!shouldSkipBR
+        && ((direction == AXDirection::Next && offset() < runs()->totalLength())
+        || (direction == AXDirection::Previous && offset()))) {
```
Could we add a comment above this if-statement stating what it's checking for? It's got a lot of conditions, which is fine — just would be nice to have plain language describing what it's doing, if possible. Or we could create named `bool` variable(s) for all / part of these conditions.

hoffmanjoshua replied:
Yeah, there are probably a few of these conditions that I can explain in a comment. I'll try to break it up + add context.

twilco reviewed (commented):
Your commit message has a trailing sentence:

"If we hit a"

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -797,6 +825,61 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
+        unsigned offset = this->offset();
+        AXTextMarker resultMarker = *this;
+
+        String flattenedRuns = currentRuns->substring(0, currentRuns->totalLength());
```
Feels like `AXTextRuns` should have a method called `toString()` or something.

```cpp
// AXTextRuns.h
struct AXTextRuns {

String substring(unsigned start, unsigned length = StringImpl::MaxLength) const;
String toString() const { return substring(0); }
};
```

hoffmanjoshua replied:
Oh yeah, let's add that

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -797,6 +825,61 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
+                // allow previous word to start at text marker if we are in the middle of the word
+                // check that end is within the word using the wordLength
+                int previousWordStart = findNextWordFromIndex(flattenedRuns, offset, false);
+                int end;
```
Do we need to re-declare `int end` here, or can we just rely on its declaration above?

hoffmanjoshua replied:
We might be able to rely on the above, let me see what happens if I remove it

hoffmanjoshua replied:
Yep, it's totally fine to remove this

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -797,6 +825,61 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
+            findWordBoundary(flattenedRuns, offset, &start, &end);
+            if (direction == AXDirection::Previous) {
+                // allow previous word to start at text marker if we are in the middle of the word
+                // check that end is within the word using the wordLength
```
WebKit coding style typically aims for "sentence cased" comments (Capital letter to start, period at the end).

Would it be more accurate to say something like:

"If `this` AXTextMarker points to the middle of a word, the previous word is..."

Not sure how to finish the sentence because I'm actually a little confused 🤔 I read your comment as saying the previous word would be this text marker? Is that right?

hoffmanjoshua replied:
This should read more like: "If this points to the middle of a word, the previous word begins at the start of that word"

hoffmanjoshua replied:
Took another look at this comment, and I don't think it makes sense anymore in the place it is at. Instead, there should be a comment above this line explaining this:
`if (start != (int)offset || end == (int)offset) {`

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -858,6 +941,53 @@ AXTextMarkerRange AXTextMarker::lineRange(LineRangeType type) const
+        startMarker = nextWordStart();
+        endMarker = startMarker.nextWordEnd();
+        auto resultRange = AXTextMarkerRange { startMarker, endMarker };
+        auto resultRangeString = resultRange.toString();
```
`AXTextMarkerRange::toString()` does a lot of work — it would be nice not to call it if we don't need to (i.e. only call it when `startMarker == *this`)

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -858,6 +941,53 @@ AXTextMarkerRange AXTextMarker::lineRange(LineRangeType type) const
+            return { WTFMove(*this), WTFMove(*this) };
+    } else {
+        startMarker = previousWordStart();
+        if (startMarker == *this)
+            return { startMarker, startMarker };
```
It could be nice to have a comment that explains why we do this — i.e. "If `this` AXTextMarker points to a word start position, the left-word is ..."

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -858,6 +941,53 @@ AXTextMarkerRange AXTextMarker::lineRange(LineRangeType type) const
+
+        endMarker = startMarker.nextWordEnd();
+        auto resultRange = AXTextMarkerRange { startMarker, endMarker };
+        auto resultRangeString = resultRange.toString();
```
Here's another spot where it could be nice to avoid unnecessarily calling `AXTextMarkerRange::toString()` — seems like we only need it when `equivalentTextPosition(endMarker)` is true.

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -858,6 +941,53 @@ AXTextMarkerRange AXTextMarker::lineRange(LineRangeType type) const
+        // There are two cases that will happen to indicate we are on the left word boundary:
+        // (1) The word break iterator returns the whitespace to the left of the word, and the end marker is the equivalent to the next starting word (handled by `isWhitespceAndEndsOnLeftWordBoundary`.
+        // (2) Thre word break iterator returns a word end before the current marker, and the next word starts at the visual equivalent to the current marker.
+        bool isWhitespaceAndEndsOnLeftWordBoundary = equivalentTextPosition(endMarker) && isASCIIWhitespace(resultRangeString[resultRangeString.length() - 1]);
```
`isASCIIWhitespace(resultRangeString[resultRangeString.length() - 1]` — do we want / need to guard against zero-length `resultRangeString` strings?

hoffmanjoshua replied:
ah yeah, I'll add a check here

twilco replied:
Just noticed a typo in this bit too: "Thre"

twilco reviewed (commented):
Thanks for your hard work on this 🙏

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -732,20 +740,38 @@ const AXTextRuns* AXTextMarker::runs() const
+    bool shouldSkipBR = ignoreBRs == IgnoreBRs::Yes && object && object->roleValue() == AccessibilityRole::LineBreak && !object->editableAncestor();
+    bool isWithinRunBounds = ((direction == AXDirection::Next && offset() < runs()->totalLength()) || (direction == AXDirection::Previous && offset()));
+    if (!shouldSkipBR && isWithinRunBounds)
+        return AXTextMarker { treeID(), objectID(), direction == AXDirection::Next ? offset() + 1 : offset() - 1 };
```
This problem existed before your PR, and can be addressed in a follow-up from either you or I...but I feel like this `offset() - 1` check is problematic for text markers with an offset of zero?

hoffmanjoshua replied:
This should be handled already. `isRunWithinBounds` prevents `offset() - 1` from happening when `offset == 0`. If we are looking at the previous direction, and `offset == 0` , then `isWithinRunBounds` will be false.

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -732,20 +740,38 @@ const AXTextRuns* AXTextMarker::runs() const
-        return { object->treeID(), object->objectID(), direction == AXDirection::Next ? 0 : object->textRuns()->lastRunLength() };
+
+        // The startingOffset is used to advance one position farther when we are coalescing object breaks and skipping positions.
+        int startingOffset = 0;
```
Can this be `unsigned`? `int` implies the possibility of a negative value, while it seems like this can only ever be 0 or 1.

hoffmanjoshua replied:
Ah good point, I'll update this

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -797,6 +823,61 @@ AXTextMarker AXTextMarker::findMarker(AXDirection direction, AXTextUnit textUnit
+                    return resultMarker;
+
+                currentRuns = currentObject->textRuns();
+                String newRunsFlattenedString = currentRuns->substring(0, currentRuns->totalLength());
```
Can we use `currentRuns->toString()` here?

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -858,6 +939,59 @@ AXTextMarkerRange AXTextMarker::lineRange(LineRangeType type) const
+            auto resultRange = AXTextMarkerRange { startMarker, endMarker };
+            auto resultRangeString = resultRange.toString();
+            // A newline should not be returned as the right word (if we are at the end of a paragraph)
+            if (resultRangeString.length() && resultRangeString[0] == '\n')
```
Is it possible to simplify this like so:

```cpp
auto rangeString = AXTextMarkerRange { startMarker, endMarker }.toString();
// A newline should not be returned as the right word (if we are at the end of a paragraph)
if (rangeString.length() && rangeString[0] == '\n')
```

hoffmanjoshua replied:
Yep, I'll change this.

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -858,6 +939,59 @@ AXTextMarkerRange AXTextMarker::lineRange(LineRangeType type) const
+            auto resultRangeString = resultRange.toString();
+            // A newline should not be returned as the right word (if we are at the end of a paragraph)
+            if (resultRangeString.length() && resultRangeString[0] == '\n')
+                return { WTFMove(*this), WTFMove(*this) };
```
You can't move something twice, and moving `*this` inside its own member function seems dangerous. Let's use copies instead:

`return  { *this, *this }`

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -858,6 +939,59 @@ AXTextMarkerRange AXTextMarker::lineRange(LineRangeType type) const
+        auto nextMarker = findMarker(AXDirection::Next, CoalesceObjectBreaks::No, IgnoreBRs::Yes);
+
+        if ((isWhitespaceAndEndsOnLeftWordBoundary || !equivalentTextPosition(endMarker)) && nextMarker == oneAheadStartMarker)
+            return { WTFMove(*this), WTFMove(*this) };
```
Same comment about moving something twice.

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -858,6 +939,59 @@ AXTextMarkerRange AXTextMarker::lineRange(LineRangeType type) const
+        bool isWhitespaceAndEndsOnLeftWordBoundary = false;
+        if (equivalentTextPosition(endMarker)) {
+            auto resultRange = AXTextMarkerRange { startMarker, endMarker };
+            auto resultRangeString = resultRange.toString();
```
Can we combine this into one line:

`auto rangeString = AXTextMarkerRange { startMarker, endMarker }.toString();`

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -858,6 +939,59 @@ AXTextMarkerRange AXTextMarker::lineRange(LineRangeType type) const
+
+bool AXTextMarker::equivalentTextPosition(const AXTextMarker& other) const
+{
+    return (findMarker(AXDirection::Next, CoalesceObjectBreaks::No, IgnoreBRs::Yes) == other || findMarker(AXDirection::Previous, CoalesceObjectBreaks::No, IgnoreBRs::Yes) == other) && objectID() != other.objectID();
```
`objectID() != other.objectID()` seems far cheaper than these `findMarker` calls. Would it make sense to move that up-front?

hoffmanjoshua replied:
Yes, we can move that up front.

twilco commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -3054,6 +3054,20 @@ static void formatForDebugger(const VisiblePositionRange& range, char* buffer, u
+        case TextUnit::RightWord:
+            return inputMarker.wordRange(WordRangeType::Right).platformData().autorelease();
+        default:
+            // TODO: Not Implemented!
```
Implemented probably doesn't need to be capitalized here.

twilco commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -3054,6 +3054,20 @@ static void formatForDebugger(const VisiblePositionRange& range, char* buffer, u
+            break;
+        }
+    }
+#endif
```
For longer compiler directive blocks like this, it's nice to prefix `#endif` with the matching `#if` condition, e.g.:

`#endif // ENABLE(AX_THREAD_TEXT_APIS)`

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -858,6 +939,59 @@ AXTextMarkerRange AXTextMarker::lineRange(LineRangeType type) const
+
+        // Check forward from the current offset for the next word start.
+        // If we are on the start of a word (next marker is a word, previous is whitespace), left word should be nil.
+        auto oneAheadStartMarker = nextWordStart();
```
When I read "one ahead", I interpret that as the `findMarker(AXDirection::Next)` from the current marker, rather than the next word start marker. Would a name like `nextWordStart` work? Or is there some reason you chose this that I'm missing?

hoffmanjoshua replied:
Yeah this should definitely include "word" in here, if not named `nextWordStart`. Will fix this

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -732,20 +740,38 @@ const AXTextRuns* AXTextMarker::runs() const
+        if ((coalesceObjectBreaks == CoalesceObjectBreaks::Yes || shouldSkipBR) && !isolatedObject()->shouldEmitNewlinesBeforeAndAfterNode())
+            startingOffset = 1;
+
+        return AXTextMarker { object->treeID(), object->objectID(), direction == AXDirection::Next ? startingOffset : object->textRuns()->lastRunLength() - startingOffset };
```
Is it possible for `object->textRuns()->lastRunLength()` to be 0 and `startingOffset` to be 1, thus making this subtraction wrap backwards?

hoffmanjoshua replied:
Hmm, this didn't seem possible from my understanding, but please correct me if I'm wrong. `object` here comes from `nextObjectWithRuns`, which only returns the next object that satisfies `hasTextRuns()`:
```c++
bool hasTextRuns() final { return textRuns().size(); }
```

But, is it possible for `AXTextRuns` to have a run with a text length of 0? If so, then yes, we should guard against that.

hoffmanjoshua replied:
Actually, this should never be possible. A few lines up, we assert that `lastRunLength` != 0:
```c++
        RELEASE_ASSERT(direction == AXDirection::Next ? object->textRuns()->runLength(0) : object->textRuns()->lastRunLength());
```

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -582,6 +584,12 @@ int AXTextMarker::lineNumberForIndex(unsigned index) const
+    auto objectRange = object->textMarkerRange();
+    unsigned textControlLength = objectRange.toString().length();
+    if (object->isTextControl() && index >= textControlLength)
+        return -1;
```
Seems like we only use the result of `AXTextMarkerRange::toString()` when `object->isTextControl()`. I think it might be more readable and more performant to do something like this?

```cpp
if (object->isTextControl() && index >= object->textMarkerRange().toString().length()) {
    // Mimic behavior of AccessibilityRenderObject::visiblePositionForIndex.
    return -1;
}
```

hoffmanjoshua replied:
Good idea!

twilco reviewed (approved):
Thanks for your hard work on this!

hoffmanjoshua committed:
AX: Implement TextUnit::LeftWord/RightWord off the main thread
https://bugs.webkit.org/show_bug.cgi?id=280825
rdar://134579687

Reviewed by Tyler Wilcock.

This patch implements the `AXLeftWordTextMarkerRangeForTextMarker` and `AXRightWordTextMarkerRangeForTextMarker` attributes using the secondary thread text markers. To mimic the live tree’s current behavior, several edges cases had to be considered:
* If at the end of a paragraph (containing block), return without a word.
* If we are finding the left word, and we are at a start word boundary, that should return the whitespace prior to the word (if any).
* If we are finding the right word, and are at the end boundary of a word, return the whitespace to the right of the word.

To verify this new behavior, the test `text-marker-word-nav` was ported over from the live tree, with some modifications. APIs that still used the live tree were removed, due to some issues with interop between live and isolated tree text markers. Three test cases were are currently failing with FIXMEs: two due to the word break iterator with Chinese characters, and one due to line break behavior.

In order to accurately accommodate handling word navigation, I had to make a change to the `findMarker` method to allow for coalescing object boundaries. This is the cases where the end position of one object is equivalent visually/in position to the start position of the next object. Coalescing mimics behavior from the live tree.

* LayoutTests/accessibility/ax-thread-text-apis/text-marker-word-nav-expected.txt: Added.
* LayoutTests/accessibility/ax-thread-text-apis/text-marker-word-nav.html: Added.
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::AXTextMarker::lineNumberForIndex const):
(WebCore::AXTextMarker::atLineBoundaryForDirection const):
(WebCore::AXTextMarker::offsetFromRoot const):
(WebCore::AXTextMarker::findLastBefore const):
(WebCore::AXTextMarker::findMarker const):
(WebCore::AXTextMarker::wordRange const):
(WebCore::AXTextMarker::equivalentTextPosition const):
(WebCore::AXTextMarker::partialOrderByTraversal const):
* Source/WebCore/accessibility/AXTextMarker.h:
(WebCore::AXTextMarker::nextWordStart const):
(WebCore::AXTextMarker::nextWordEnd const):
(WebCore::AXTextMarker::previousWordStart const):
(WebCore::AXTextMarker::previousWordEnd const):
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRuns::toString const):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
(WebCore::AccessibilityRenderObject::textRuns):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper textMarkerRangeAtTextMarker:forUnit:]):

Canonical link: https://commits.webkit.org/284938@main
